File: blackjack/golang/exact.go
package main

import "fmt"
import	"github.com/BenLubar/memoize"

/* Decks */

const decks = 1

const hit_soft_17 = false

var dealer_p func(total int,ace bool,cards [10]int,n int) ([22]float64)

func init() {

	dealer_p = memoize.Memoize(func(total int,ace bool,cards [10]int,n int) ([22]float64) {

	var outcomes [22]float64

	switch {
	case total > 21:
		outcomes[0] = 1.0
	case total >= 17:
		outcomes[total] = 1.0
	case (total==7) && ace && !(hit_soft_17):
		outcomes[total+10] = 1.0
	case (8 <= total && total <= 11) && ace:
		outcomes[total+10] = 1.0
	default:
		/* Hit */
		for i := 0; i < 10; i++ {
			if (cards[i] > 0) {
				p := float64(cards[i])/float64(n)
				cards[i] += -1
				d := dealer_p(total+i+1,(ace || i==0),cards,n-1)
				for j := 0; j < 22; j++ {
					outcomes[j] += p*d[j]
				}
				cards[i] += 1
			}
		}

	}

	return(outcomes)
			
	}).(func(total int,ace bool,cards [10]int,n int) ([22]float64))

}

var player_p func(total int,ace bool,cards [10]int,dc int,n int) (float64,float64,float64)

func init() {

	player_p = memoize.Memoize(func(total int,ace bool,cards [10]int,dc int,n int) (float64,float64,float64) {

		var e_stand float64 = 0.0
		var e_hit float64 = 0.0
		var e_opt float64 = 0.0

		if total <= 11 && ace {
			d := dealer_p(dc,(dc==0),cards,n)
			for i := 0; i < total+10; i++ {
				e_stand += d[i]
			}
			for i := total+11; i < 22; i++ {
				e_stand += -d[i]
			}
		} else {
			d := dealer_p(dc,(dc==0),cards,n)
			for i := 0; i < total; i++ {
				e_stand += d[i]
			}
			for i := total+1; i < 22; i++ {
				e_stand += -d[i]
			}
		}

		/* Hit */

		for i := 0; i < 10; i++ {
			if cards[i]>0 {
				p := float64(cards[i])/float64(n)
				if total+i+1 > 21 {
					e_hit += -p
				} else {
					cards[i] += -1
					e_o,e_s,e_h := player_p(total+i+1,(ace || i==0),cards,dc,n-1)
					_ = e_s
					_ = e_h
					e_hit += p*e_o
					cards[i] += 1
				}
			}
		}
		if (e_hit >= e_stand) {
			e_opt = e_hit
		} else {
			e_opt = e_stand
		}
		return e_opt,e_hit,e_stand
		
	}).(func(total int,ace bool,cards [10]int,dc int,n int) (float64,float64,float64))
}
		
func main() {
	var deck [10]int
	var p [10]float64

	for i := 0; i < 9; i++ {
		deck[i] = decks*4
		p[i] = 4.0/52
	}
	deck[9] = decks*16
	p[9] = 16.0/52

	for i := 1; i < 9; i++ {
		deck[i] += -1

		for j := 0; j < 10; j++ {
			deck[j] += -1
			for k := 0; k < 10; k++ {
            
				deck[k] += -1
				e_o,e_h,e_s := player_p(j+k+2,(j==0 || k==0),deck,i+1,(decks*52-3))
/*				if (e_h>e_s) {
					s := "hit"
				} else {
					s := "stand"
				}*/
				_ = e_o
				/*_ = s*/
				fmt.Println(i+1,j+1,k+1,e_h,e_s)   
				deck[k] += 1
			}
			deck[j] += 1
		}
		deck[i] += 1
	}

}

File: blackjack/python/exact.py
#!/usr/bin/env python3

#import numpy as np

# Number of decks

n = 1

deck = [4*n]*9
deck.append(16*n)

# Dealer hits soft 17
hit_soft_17 = False

# Double after split
double_after_split = True

# Resplit aces
resplit_aces = False

class memoize:
    def __init__(self, func):
        self.func = func
        self.known = {}

    def __call__(self, *args, **kwargs):
        key = give_fake_hash((args, kwargs))

        try:
            return self.known[key]
        except KeyError:
            value = self.func(*args, **kwargs)
            self.known[key] = value
            return value

def give_fake_hash(obj):
    cls = type(obj)
    name = "Hashable" + cls.__name__

    def fake_hash(self):
        return hash(repr(self))

    t = type(name, (cls, ), {"__hash__": fake_hash})

    return t(obj)

@memoize
def dealer_p(total, ace, cards, n):

    outcomes = [0.0]*22
    
    if (total > 21):
        
        # Dealer busts
        outcomes[0] = 1.0
        
    elif (total >= 17):

        # Dealer stands
        outcomes[total] = 1.0

    elif ((total==7) and ace and not(hit_soft_17)):
        
        # Dealer stands on a soft 17
        outcomes[total+10] = 1.0

    elif ((8 <= total <= 11) and ace):

        # Dealer stands on a soft total > 17
        outcomes[total+10] = 1.0

    else:

        # Dealer hits

        for i in range(0,10):
            if (cards[i]>0):
                p = cards[i]/n
                cards[i] += -1
                # Can we speed this up?
                d = dealer_p(total+i+1, (ace or i==0), cards, n-1)
                for j in range(0,22):
                    outcomes[j] += p*d[j]
                #l = [p*x for x in dealer_p(total+i+1,(ace or i==0),cards,n-1)]
                #outcomes[:] = [a+b for a, b in zip(outcomes, l)]
                cards[i] += 1

    return(outcomes)

@memoize
def player_p(total, ace, cards, dc, n):

    # Bust
    if (total > 21):
        return(-1.0, 'bust')

    # Stand
    if (total <= 11 and ace):
        # Stand on total+10
        d = dealer_p(dc, (dc==0), cards, n)
        e_stand = sum(d[0:total+10])-sum(d[total+11:])
    else:
        # Stand on total
        d = dealer_p(dc, (dc==0), cards, n)
        e_stand = sum(d[0:total])-sum(d[total+1:])

    # Hit
    e_hit = 0.0
    for i in range(0,10):
        if (cards[i]>0):
            p = cards[i]/n
            cards[i] += -1
            e_card, strategy = player_p(total+i+1,(ace or i==0),cards,dc,n-1)
            e_hit += p*e_card
            cards[i] += 1

    # If we haven't busted we can always stand
    e_best = e_stand
    strategy = 'stand'
    if (e_hit > e_best):
        e_best = e_hit
        strategy = 'hit'
    return(e_best, strategy)

for i in range(1, 9):
    # Dealer
    deck[i] += -1

    for j in range(0,10):
        deck[j] += -1
        for k in range(j,10):
            
            deck[k] += -1
            e, s = player_p(j+k+2, (j==0 or k==0), deck, i+1, sum(deck))
            print('Dealer showing {}, hand {}/{} : {}, E = {:4.2f}'.format(i+1,j+1,k+1,s,e))
            deck[k] += 1
        deck[j] += 1
    deck[i] += 1
